### PR TITLE
CORDA-1213 - Explicitly disallow serialization of non static nested classes

### PR DESCRIPTION
**Pull request by fenryka**
Also run the auto reformatter

### CODE REVIEW

File: docs/source/changelog.rst [hunk 1]
```diff
@@ -7,6 +7,12 @@ Unreleased
 Here are brief summaries of what's changed between each snapshot release. This includes guidance on how to upgrade code
 from the previous milestone release.
 
+* Serializing a non static nested class (an ``inner class`` in Kotlin) will be rejected explicitly by the serialization
```
Comment (rick-r3) [inline docs/source/changelog.rst]: Also an inner class in Java, it's just implicit, and requires `static` to make it static and no longer an inner class.

I think it is better to describe them as inner classes with a (strictly, a non-static nested class in JVM parlance) clarifier.

You could also say that if they wish to have something like this, it's better to have explicit constructor arg on a **static** nested class.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt [hunk 2]
```diff
@@ -58,13 +57,22 @@ open class ObjectSerializer(val clazz: Type, factory: SerializerFactory) : AMQPS
...
             type: Type,
             output: SerializationOutput,
-            debugIndent: Int) = ifThrowsAppend({ clazz.typeName }) {
+            debugIndent: Int) = ifThrowsAppend({ clazz.typeName }
+    ) {
+        if (propertySerializers.size != javaConstructor?.parameterCount &&
+                javaConstructor?.parameterCount ?: 0 > 0
+        ) {
+            throw NotSerializableException("Serialization constructor for class $type expects "
+                    + "${javaConstructor?.parameterCount} parameter but we have ${propertySerializers.size} "
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: parameter -> parameters

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt [hunk 3]
```diff
@@ -58,13 +57,22 @@ open class ObjectSerializer(val clazz: Type, factory: SerializerFactory) : AMQPS
...
+                    + "properties to serialize")
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: full stop

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt [hunk 4]
```diff
@@ -136,7 +144,12 @@ open class ObjectSerializer(val clazz: Type, factory: SerializerFactory) : AMQPS
     fun construct(properties: List<Any?>): Any {
         logger.trace { "Calling constructor: '$javaConstructor' with properties '$properties'" }
 
-        return javaConstructor?.newInstance(*properties.toTypedArray()) ?:
-                throw NotSerializableException("Attempt to deserialize an interface: $clazz. Serialized form is invalid.")
+        if (properties.size != javaConstructor?.parameterCount) {
+            throw NotSerializableException("Serialization constructor for class $type expects "
+                    + "${javaConstructor?.parameterCount} parameter but we have ${properties.size} serialized properties")
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: parameter -> parameters

Comment (rick-r3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: full stop

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 5]
```diff
@@ -158,7 +160,7 @@ fun Class<out Any?>.propertyDescriptors(): Map<String, PropertyDescriptor> {
             PropertyDescriptorsRegex.re.find(func.name)?.apply {
                 // matching means we have an func getX where the property could be x or X
                 // so having pre-loaded all of the properties we try to match to either case. If that
-                // fails the getter doesn't refer to a property directly, but may to a cosntructor
+                // fails the getter doesn't refer to a property directly, but may to a constructor
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: may -> may refer

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 6]
```diff
@@ -219,19 +221,26 @@ internal fun <T : Any> propertiesForSerializationFromConstructor(
 
     val classProperties = clazz.propertyDescriptors()
 
+    // Annoyingly there isn't a better way to ascertain that the constructor for the class
+    // has a synthetic parameter inserted to capture the reference to the outer class. You'd
+    // think you could inspect the parameter and check the isSynthetic flag but that is always
+    // false so given the naming convention is specified by the standard we can just check for
+    // this
+    if (kotlinConstructor.javaConstructor?.parameterCount ?: 0 > 0 &&
+            kotlinConstructor.javaConstructor?.parameters?.get(0)?.name == "this$0"
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: Does this hold if it's a inner class inside an inner class for example?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt (hunk 2 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt (hunk 3 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt (hunk 4 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt (hunk 6 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: yep, added tests :)

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt (hunk 5 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: done

**Comment by fenryka** [comment]
@rick-r3 back to you

File: docs/source/changelog.rst [hunk 7]
```diff
@@ -7,6 +7,12 @@ Unreleased
 Here are brief summaries of what's changed between each snapshot release. This includes guidance on how to upgrade code
 from the previous milestone release.
 
+* Serializing an inner class (non-static nested class in Java, inner class in Kotlin) will be rejected explicitly by the serialization
+  framework. Prior to this change it didn't work, but the error thrown was opaque (complaining about too few arguments
+  to a constructor). Whilst this was possible in the older Kryo implementation (Kryo passing null as the synthesised
+  reference to the outer class) as per the Java documentation `here <https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html>`_
+  we are disallowing this as the paradigm in general makes little sense for Contract States
```
Comment (rick-r3) [inline docs/source/changelog.rst]: full stop

**Pull request merged**